src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs

```diff
@@ -699,101 +826,71 @@ public SortKey[] SortKeys
                     }
                 }
 
+                _keysAsnLength = 0;
                 _keys = new SortKey[value.Length];
                 for (int i = 0; i < value.Length; i++)
                 {
                     _keys[i] = new SortKey(value[i].AttributeName, value[i].MatchingRule, value[i].ReverseOrder);
+                    _keysAsnLength += 13 + (value[i].AttributeName?.Length ?? 0) + (value[i].MatchingRule?.Length ?? 0);
```

Comment:
I strongly advise ditching all of the size estimation and just using a shared writer with the default 1k first page.